pkg/controller/release/strategy_executor.go

```diff
@@ -147,7 +149,30 @@ func (e *StrategyExecutor) Execute(prev, curr, succ *releaseInfo) (bool, []Strat
 	pipeline := NewPipeline()
 	pipeline.Enqueue(genInstallationEnforcer(ctx, curr, succ))
 
-	if isHead {
+	// when a release is steppingBackwards, the strategy executor needs to reverse it's direction.
+	// a release is "not steppingBackwards" when the targetStep >= achieved step.
+	// we include the state where the steps are equal because this state does not require
+	// reversing the direction of the strategy executor
+	if e.steppingBackwards {
+		// release is steppingBackwards (achieved step > target step):
+		// increase capacity for previous release
+		// increase traffic for previous release
+		// reduce traffic for current release
+		// reduce capacity for current release
+		if isHead {
+			if hasTail {
+				prevctx := ctx.Copy()
+				prevctx.isHead = false
+				pipeline.Enqueue(genCapacityEnforcer(prevctx, prev, curr))
+				pipeline.Enqueue(genTrafficEnforcer(prevctx, prev, curr))
```

Comment:
This is definitely out of the scope of this MR, but for the future: I find this snippet quite confusing. As far as I understand, the order of Capacity and Traffic enforcing is something that depends entirely on whether you are "increasing" or "decreasing". If that is the case and there's no other case I'm ignoring here, I feel an abstraction of the kind "increaseForRelease" or something else aptly named that takes care of the correct order of enforcements would contribute to clarity here. Otherwise this can be tricky to read and can potentially lead to mistakes when working on this part.

Reply:
You are seeing one of the reasons I would like to make changes to this code. I agree with you, if we look at this as a state machine, we should be able to construct the notion of "increasing" or "decreasing" for release. However, the strategy executor is not implemented as a state machine, but as a pipeline.

Reply:
Ah, right, that would definitely be ideal - but even without that, could we not have another method that enqueues the actions to the pipeline in the correct order depending on whether we are increasing or decreasing?

Reply:
I don't think I fully understand what you mean.
I think somewhere something in shipper needs to know the direction of the progress of a given release, in order to do this in the correct order. 
Generally speaking, Shipper should be able to look at a release and know what exactly to enforce on it. All the data about this release (if it's the contender, if it's increasing or decreasing, and if it needs capacity or traffic enforcing) is already in the body of the release (spec and status)

Reply:
Hmm, I'm probably misunderstanding something. What I meant was literally:
```
func increase(ctx, rel, rel) {
  pipeline.Enqueue(genCapacityEnforcer(ctx, rel, rel))
  pipeline.Enqueue(genTrafficEnforcer(ctx, rel, rel))
}
func decrease(ctx, rel, rel) {
  pipeline.Enqueue(genTrafficEnforcer(ctx, rel, rel))
  pipeline.Enqueue(genCapacityEnforcer(ctx, rel, rel))
}
```

Reply:
Ah ok now I understand what you mean.
Yeah that might make things easier to read.

Reply:
Sorry, I didn't explain it really clearly at the beginning 😅